### PR TITLE
Using random generator instead of character shuffle for data obfuscation

### PR DESCRIPTION
**Pull request by compwron**
As per https://twitter.com/nettofarah/status/670757721851883520

### CODE REVIEW

**Comment by nettofarah** [comment]
hey, @compwron.
Thank you for submitting this PR.

I'm curious as to what's the motivation behind choosing `SecureRandom.urlsafe_base64` as a strategy for this.

The reason I chose to split the characters and scramble them was so we could keep column types consistent. (i.e. reduce the chances of breaking database type constraints).

Another benefit of Polo's current approach is to keep the record data consistent to `ActiveRecord` validations. Not that there's any guarantee that the model's validation is still going to pass, but there's a higher chance.

Your implementation will probably break for numeric and short columns. (e.g. `INTEGER` or `VARCHAR(5)`).

Here's a couple of ideas:

a) Detect the data type you're trying to randomize and apply a specific strategy to guarantee data consistency. (probably not so easy to implement)

b) Allow users to pass in their own obfuscation strategy for a given field (this could be a lambda or a ruby class). That way you could implement custom obfuscation strategies for different fields.
For example, you could pass in a lambda block that makes sure an `email` field is still a valid email.

```ruby
lambda { |field| "#{SecureRandom.urlsafe_base64}@example.com" }
```

**Comment by nettofarah** [comment]
The setup could look like this:

```ruby
Polo.configure do
  obfuscate :email, strategy: lambda do |field|
    "#{SecureRandom.urlsafe_base64}@example.com"
  end
end

```

**Comment by compwron** [comment]
Hmm, good point. I will take a look. I make no promises about timeliness.

**Comment by compwron** [comment]
How does this solution look? I believe that this preserves current behavior, while still allowing for more complex obscuring strategies to be passed in.

**Comment by nettofarah** [comment]
Yeah! This is what I had in mind.
I'll take another look tomorrow just to make sure, but this is looking good so far.

File: lib/polo/translator.rb [hunk 1]
```diff
@@ -35,12 +35,24 @@ def instances
...
+          instance.send("#{field}=", new_field_value(field, function, value))
         end
       end
     end
+
+    def intersection(attrs, fields)
+      attrs & fields.to_a.flatten.map(&:to_s)
+    end
+
+    def new_field_value(field, function, value)
```
Comment (nettofarah) [inline lib/polo/translator.rb]: maybe rename `function` with `strategy`

File: spec/translator_spec.rb [hunk 2]
```diff
@@ -15,12 +16,35 @@
...
-        translator = Polo::Translator.new([finder], Polo::Configuration.new(obfuscate: [:email]))
-        netto = translator.instances.first
+      let(:translator) { Polo::Translator.new([finder], Polo::Configuration.new(obfuscate: obfuscated_fields)) }
+      let(:netto) { netto = translator.instances.first }
 
-        expect(netto.email).to_not be_nil
-        expect(netto.email).to_not eq('nettofarah@gmail.com')
+      context "with obfuscated field with no specified strategy" do
+        let(:obfuscated_fields) {{email: nil }}
+        it "shuffles characters in field" do
```
Comment (nettofarah) [inline spec/translator_spec.rb]: let's add a space between `let` and `it`.

File: spec/translator_spec.rb [hunk 3]
```diff
@@ -15,12 +16,35 @@
...
+          expect(netto.email).to_not be_nil
+          expect(netto.email.length).to eq email.length
+          expect(sorted_characters(netto.email)).to eq sorted_characters(email)
+        end
+      end
+
+      context "with obfuscated field with integer strategy" do
+        let(:obfuscated_fields) {{email: lambda { |_| 42 } } }
```
Comment (nettofarah) [inline spec/translator_spec.rb]: let's add a space between `let` and `it`.

Comment (nettofarah) [inline spec/translator_spec.rb]: Maybe we make this a little more clear here.
I get what's going one, but maybe "integer strategy" is too generic.

File: spec/translator_spec.rb [hunk 4]
```diff
@@ -15,12 +16,35 @@
...
+        it "replaces contents of field according to the supplied lambda" do
+          expect(netto.email.to_s).to eq "42"
+        end
+      end
+
+      context "with no strategy passed in" do
+        let(:obfuscated_fields) { [:email] }
```
Comment (nettofarah) [inline spec/translator_spec.rb]: let's add a space between `let` and `it`.

**Comment by nettofarah** [comment]
@compwron: this looks fantastic.
Just a few things before this PR is good to merge:

1) I left a few comments around styling, just to make the PR consistent with the rest of the codebase.

2) We need to squash all the commits keep history clean. We're trying to follow this standard: http://chris.beams.io/posts/git-commit/
in short: Short and concise title, followed by a paragraph explaining the motivation/change behind that feature.

3) Add some documentation to the README explaining what this feature does.
This the first major feature after launch, so we need to make sure people know about this.

Thank you!

**Comment by compwron** [comment]
I tried. @nettofarah what do you think?

**Comment by nettofarah** [comment]
This looks fantastic!
Thank you and congrats on the awesome pull request.

**Comment by compwron** [comment]
:)

**Pull request merged**